### PR TITLE
Fix mishandling of sigma in guess_elim (regression from 8.11)

### LINKED ISSUE #11722: Spurious universes declared when doing destruct
#### Description of the problem

Coq generates useless universes when doing induction/destruct in universe polymorphic mode,
problem introduced in commit a9e4f78935ba091d7e012df8398bccc5d08d1370

```coq
Require Import Program.
Set Universe Polymorphism.

Inductive paths@{i} (A : Type@{i}) (a : A) : A -> Type@{i} :=
  idpath : paths A a a.

Inductive nat :=
  | O : nat
  | S : nat -> nat.

Axiom cheat : forall {A}, A.

Program Definition foo@{i} : forall x : nat, paths@{i} nat x x := _.
Next Obligation.
  destruct x.
  constructor.
  apply cheat.
Defined. (* Universe unbound error *)

Check foo@{_}.
```

#### Coq Version

Only master is impacted.

### PR DESCRIPTION
**Pull request by mattam82**
In case no eliminator is given, we wait until `get_eliminator` to
produce the actual eliminator. Using the sigma produced by guess_elim
here would insert an unused universe for the predicate's type in the
sigma.

**Kind:** bug fix

Should have no impact on user developments (except maybe HoTT-like developments that use destruct/induction) which will have less useless universes. Let's see on HoTT.

Fixes / closes #11722 

- [x] Added / updated test-suite

### CODE REVIEW

**Comment by mattam82** [comment]
It's more in @SkySkimmer 's ballpark as he introduced the problem recently

**Comment by mattam82** [comment]
This bug prevents Equations for HoTT on master to not produce bloated definition w.r.t. universes.

**Comment by SkySkimmer** [comment]
Isn't the real issue that there's a missing restrict in the obligation system? eg 
~~~coq
Require Import Program.Tactics.

Set Universe Polymorphism.
(* needed because we don't enforce universe declarations on
   obligations, in univ poly mode however the universes must be
   redeclared for the main definition so we do see the error *)

Program Definition bla@{} : nat := _.
Next Obligation.
  let x := constr:(Type) in exact 0.
Defined. (* unbound univ *)
~~~

**Comment by mattam82** [comment]
Both would be nice, but there's no point in declaring this useless universe.

File: tactics/tactics.ml [hunk 1]
```diff
@@ -4250,10 +4250,11 @@ let find_induction_type isrec elim hyp0 gl =
...
        let sort = Tacticals.New.elimination_sort_of_goal gl in
-       let sigma, (elimc,elimt),_ = guess_elim isrec false sort hyp0 gl in
-       let scheme = compute_elim_sig sigma ~elimc elimt in
-       (* We drop the scheme waiting to know if it is dependent *)
-       sigma, scheme, ElimOver (isrec,hyp0)
+       let sigma', (elimc,elimt),_ = guess_elim isrec false sort hyp0 gl in
+       let scheme = compute_elim_sig sigma' ~elimc elimt in
+       (* We drop the scheme waiting to know if it is dependent, this
+          needs no update to sigma at this point. *)
+       Tacmach.New.project gl, scheme, ElimOver (isrec,hyp0)
```
Comment (ppedrot) [inline tactics/tactics.ml]: This is correct, but the way to code is written doesn't make it obvious that the newer evarmap is never going to be referred to in the return branches. I wouldn't mind marking this a in need of a TODO.

Comment (mattam82) [reply tactics/tactics.ml]: Why not, we could also just return the indref and nparams instead of the scheme which might otherwise depend on sigma'

Comment (mattam82) [reply tactics/tactics.ml]: Done

**Review by ppedrot — APPROVED** [review]

**Comment by SkySkimmer** [comment]
I'll let someone else merge this (@ppedrot ?)

**Pull request merged**